### ISSUE TITLE
Envelope Following via parameter Modulation

### ISSUE THREAD

**Issue by Justinmac84** [nvaccess/osara#71]
Hi, 

When an effect is selected in the FX chain dialogue and you press the "param" button, go to "fx parameter list"/"parameter modulation" and select a parameter, there is the option to control the value of that parameter with an audio control signal, i.e. input from another track or channel,  as a more advanced and flexible form of side-chaining. 

In order to do this, you need to set the baseline slider, i.e. the starting value for the parameter from which it will respond to the control signal. 

This slider is currently pretty much inaccessible. There is a way to make it move, which makes NVDA's progress bar beep trigger, however this has been found to be temperamental across different builds of Reaper - possibly portable versus full installation, possibly 32-bit versus 64-bit. 

Would it be possible to make this slider easier to focus and have speech report the slider's position? 

the implications of such a thing being made accessible, assuming it's even possible are wide-ranging but here are a few. 

If you have a compressor or noise gate you particularly like that doesn't support side-chaining, using this method you can make it happen.

If you want an eq and a compressor effect to work together, which can be useful in some mixing situations, as described in Reaper's documentation, I forget whether it's the wiki or the manual, this is how to do it. 

If you want the panning of one track to follow the panning of another, this is how to do it. 

There are also a tremendous number of sound design implications such as the ability to create level modulated reverbs or filters etc. 

thanks and I hope this explains everything, 

Justin

**Comment by GianlucaApollaro** [comment]
I could make this slider move only in my installed reaper. No matter what I try in any other portable either 32 or 64 bit versions of reaper. It works only in my installed reaper that is latest version with latest jaws/NVDA and latest osara

**Comment by jcsteh** [comment]
I can focus that slider easily enough with NVDA object navigation. It still isn't overly accessible--none of those custom REAPER sliders are--but I can move the slider with the arrow keys and the "output" gauge changes value, which explains the progress beeps. Note that it only shows a value if one of the other check boxes are checked. In fact, I think one of the other check boxes has to be checked for this to work at all.

That said, I'm not sure what scale the values here use. Hopefully, it's the same as other controls like this.

Reporting the values depends on #54.

**Comment by GianlucaApollaro** [comment]
Hi,
I can move the slider only on an installed version of reaper. No matter 
what I try it does not move on any portable reaper
I can't explain this.

Il 22/02/2016 12:57, James Teh ha scritto:

**Comment by jcsteh** [comment]
Just tested on a portable and it works fine here. As I say, one of the 
other check boxes must be checked in order for the output gauge to move; 
e.g. Audio control signal (sidechain). You'll need to disable NVDA 
simple review, tab to the baseline check box, move to containing object, 
move to next object (which should be a window) and move focus to 
navigator. The key point is that you have to focus the window containing 
the slider, not the slider itself, because REAPER's accessibility 
implementation is buggy.

**Comment by GianlucaApollaro** [comment]
I'll try what you suggest, but, why, according to you can I move the 
slider only on the installed version? I perform the same steps on both 
versions:
1. Check the sidechain box
2. Tab once to get to the baseline checkbox
3. Using up or down arrow keys to move the slider.
4. At this point I get it moving only on the installed reaper.
So does this mean REAPER's accessibility
implementation is buggy only if you use a portable version?

Thanks for your suggestions

Il 22/02/2016 13:12, James Teh ha scritto:

**Comment by Justinmac84** [comment]
Hi,

I read all number 54 but won't pretend to understand it all.

Would it be possible at least to create actions that move the baseline 
slider in small increments, which could then be bound to shortcut keys 
if creating a hack or custom interface isn't possible? Even more 
importantly, would it be possible to create an action to place the 
baseline value in the centre? For many plugs such as EQs and panning 
effects, this represents unity gain and it would be important to get 
right on the nose, which i don't believe would be possible with the 
progress beeps, even assuming you weren't using a screen reader other 
than NVDA.

It is curious that, even if object nav works for both full and portable 
installs of Reaper, it is definitely easier with the full install, in 
which you can just tab to the baseline checkbox and use the arrow keys 
to get the beeps. This definitely doesn't work in the portable, even 
with all the right boxes checked,

Justin

On 22/02/2016 11:57, James Teh wrote:
>
> I can focus that slider easily enough with NVDA object navigation. It 
> still isn't overly accessible--none of those custom REAPER sliders 
> are--but I can move the slider with the arrow keys and the "output" 
> gauge changes value, which explains the progress beeps. Note that it 
> only shows a value if one of the other check boxes are checked. In 
> fact, I think one of the other check boxes has to be checked for this 
> to work at all.
>
> That said, I'm not sure what scale the values here use. Hopefully, 
> it's the same as othe
>

I did read issue number 54, but won't pretend to have understood all of it.

Would it be possible to create actions that move the values of the 
baseline slider up and down so that they could be bound to a shortcut key?
It is curious that, even though object nav may well work on both 
portable and full installs of ReaIn that, you don't need object nav at 
all, you can just tab to the baseline checkbox, which is checked by 
default, and move the cursor up and down. That definitely doesn't work 
with the portable version,

Justin per, with a full install you don't need object nav at all, just 
to tab to the checkbox and then use the arrow keys.

I'm pretty sure this method doesn't work in portable as I have tried it 
with all the right checkboxes checked.


Even more importantly, would it be possible to create an action or 
define some way of putting the baseline slider in the centre? For some 
plug-ins, this represents unity gain and it would be vital to get this 
exactly right. Without perfect pitch, that wouldn't be possible just by 
using the progress beeps as a guide would it and, even if it would, 
would it be accessible to all screen readers?


> Hi,
>

Thanks very much for following this up.

**Comment by jcsteh** [comment]
For me, it makes no difference whether I use a portable or an installed 
copy. I'm not sure why it would for you.

I did just discover something else, though. I was pretty sure that 
pressing the arrow keys while the check box was focused wouldn't adjust 
the slider. For me, at least, it seems I'm right. However, if you press 
down arrow, the focus moves to the slider, but you don't know about it 
due to the REAPER accessibility bugginess. In short, tab to the baseline 
check box, press down arrow, and then you can adjust the slider. I'm 
wondering whether perhaps you just happened to hit down arrow first with 
the installed copy but not the portable.

**Comment by GianlucaApollaro** [comment]
Hi,
I also discovered the down arrow trick, but it works only in the 
installed version. I couldn't get the object navigation to work either 
in the portable.

Il 22/02/2016 23:41, James Teh ha scritto:

**Comment by Justinmac84** [comment]
When you say move focus to navigator, do you mean by pressing nvda 
numpad minus or am I pressing the wrong key? If I'm pressing the right 
one, pressing that followed by the arrow keys still makes no beeps for me,

Justin

**Comment by jcsteh** [comment]
Well, now that we've discovered down arrow from the check box gets focus 
to the slider, you can ignore the object navigation bit. :)

**Comment by Justinmac84** [comment]
Except that it definitely does not work on my portable version of 
Reaper, whether or not I start with cursor down or cursor up, nor 
Genluka's, nor his friends. You must have the magic touch,

Justin

**Comment by ScottChesworth** [comment]
Just FYI, I'm throwing in one more report of being able to reproduce
these swanky moves working on an installed copy, but no dice with a
portable Reaper.

**Comment by Justinmac84** [comment]
In any case, in order for the feature to be usable, we'd have to be able 
to find 50% exactly at the very least and knowing values with greater 
precision would be ideal,

Justin

**Comment by jcsteh** [comment]
Going to see if I can get any contact with Cockos about this 
accessibility implementation. Trying to hack around stuff like this is 
getting ridiculous.

**Comment by Justinmac84** [comment]
I did file an issue about this, in layman speak in the feature request 
forum. Perhaps, if that's the best way, you could bulk it out with the 
tech speak to ilucidate my vagueness.

Justin

**Comment by ScottChesworth** [comment]
Vic's had the most recent consistent contact with them, so far as I know. He's probably your man to make a CC introduction.

**Comment by jcsteh** [comment]
After technical discussion with Cockos, this is fixed in REAPER 5.20pre13.

**Comment by Justinmac84** [comment]
You already closed the issue so I hope this won't reopen it or anything 
but thanks a lot, this is awesome. Really appreciate you reaching out to 
them,

Justin

**Comment by GianlucaApollaro** [comment]
This is a good news. Even though I tried a portable of the last preview 
and although the slider doesn't move with up or down arrow, I can put it 
to 50% by left clicking on it.

Il 27/02/2016 22:25, Justinmac84 ha scritto:

**Comment by jcsteh** [comment]
That is so bizarre that it doesn't move for you. I can't help thinking there must be a setting different between the two copies. What if you literally copy your installed config into your portable; i.e. create a portable, then just copy %appdata%\reaper into it?
